### PR TITLE
feat(prometheus): expose tags as labels

### PR DESCRIPTION
**Pull request by carnei-ro**
### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [x] [There is a user-facing docs PR against](https://github.com/Kong/docs.konghq.com/pull/5120) https://github.com/Kong/docs.konghq.com

### Full changelog

* Prometheus plugin now allows exposing Service, Route and/or Consumer tags as metric labels.

### CODE REVIEW

File: kong/plugins/prometheus/handler.lua
Comment (colinh-kong) [inline kong/plugins/prometheus/handler.lua:44]: My lua is rusty but could these not be written using `table.insert` such as

```suggestion
    for _, tag in pairs(message.service.tags) do
      table.insert(tags, tag)
    end
```

**Review by colinh-kong — CHANGES_REQUESTED** [review]
I haven't tested my code change suggestions so don't accept them without verifying first ;)

Does this have an early docs PR?

File: kong/plugins/prometheus/handler.lua
Comment (colinh-kong) [inline kong/plugins/prometheus/handler.lua:58]:
```suggestion
  table.sort(tags)
  serialized.tags = table.concat(tags, ",")
```

File: kong/plugins/prometheus/handler.lua
Comment (carnei-ro) [reply kong/plugins/prometheus/handler.lua:44]: The idea here is: do not duplicate strings; `table.insert` will insert the string as the "table value" and the index will be `index++`;

File: kong/plugins/prometheus/handler.lua
Comment (carnei-ro) [reply kong/plugins/prometheus/handler.lua:58]: this code will collect all the keys from the table (that does not have duplicated strings, as they are the keys), sort it (to avoid any troubles with cardinality) then concatenate it

File: kong/plugins/prometheus/handler.lua
Comment (carnei-ro) [reply kong/plugins/prometheus/handler.lua:44]: example of the result for your suggestion: https://www.jdoodle.com/iembed/v0/C2b

Comment (colinh-kong) [reply kong/plugins/prometheus/handler.lua:44]: Ah alright that is not at all how I expected `table.insert` to operate :+1:

**Comment by carnei-ro** [comment]
> I haven't tested my code change suggestions so don't accept them without verifying first ;)
> 
> Does this have an early docs PR?

About the early docs: I'm still figuring out what I have to do there.; There are no docs about Kong 3.x

**Comment by colinh-kong** [comment]
The prometheus plugin [CHANGELOG](https://docs.konghq.com/hub/kong-inc/prometheus/#changelog) section comes to mind re: docs(?)

**Comment by carnei-ro** [comment]
@colinh-kong rebased and created the PR https://github.com/Kong/docs.konghq.com/pull/5120

File: kong/plugins/prometheus/handler.lua
Comment (hbagdi) [inline kong/plugins/prometheus/handler.lua:59]: This plugin has already faced a lot of performance issues under high request volume and high cardinality of services/routes.
I'm a bit concerned about the performance impact of this change.
We will need to measure performance before we hit the merge button.

**Comment by hbagdi** [comment]
Closing because of lack of activity. We would like to accept this contribution if anyone wishes to pick this back up.

**Pull request closed without merging**
